Timer events

Boosik opened the pull request:
Timer event starts a timer which will fail the quest if a player not finish it in time.
Cancel timer event cancels the timer so it will not fail the quest if the stage is finished.
All players timers are canceled at logout (and related quests will fail).

Also fixed Finish stage events, which were not firing at all.

Boosik commented:
I forget to remove timers map entry on task finish. Will fix it tomorrow morning.

Boosik commented:
Fixed possible memory leak (properly remove finished tasks from timer map).

FlyingPikachu commented on src/main/java/me/blackvein/quests/ObjectiveTimer.java:
```diff
@@ -0,0 +1,33 @@
+    Quests plugin;
+    Quest quest;
+    Integer time;
+    Boolean last;
```
Why use class rather than primitive type? (Integer rather than int?)

Boosik replied:
This one has no reason, I'll change it.

FlyingPikachu commented on src/main/resources/quests.yml:
```diff
@@ -76,4 +76,25 @@ quests:
+          finish-event: CancelTimer
+        '2':
+          password-displays:
+          - "Do you want to finish? (Type yes before the time runs out!)"
```
Why should the user need to confirm they've finished? I can imagine the frustration of not being able to type "yes" with only a second remaining.

Boosik replied:
Normally user doesn't need to confirm finish. It is not required for timers to work correctly. I originally added the second stage to test out if the CancelTimer event fires and works properly.

FlyingPikachu commented on src/main/java/me/blackvein/quests/Event.java:
```diff
@@ -232,13 +241,51 @@ public void fire(Quester quester, Quest quest) {
+		}
+		if (cancelTimer) {
+			quester.timers.entrySet().stream()
+					.filter(questEntry -> Objects.equals(questEntry.getValue().getName(), quest.getName()))
```
Would like to avoid Java 8 dependency if possible (the "->" lambda expression). This is because Quests supports older versions of the Minecraft server which will not necessarily be run with Java 8.

Boosik replied:
Seems reasonable. But Java 8 target is specified in pom, wouldn't it cause major-minor exception regardless of my use of streams and lambdas if run on older Java version?

FlyingPikachu replied:
Quite possibly, in which case I'll want to correct that. Thanks!

FlyingPikachu commented:
First, thanks for your PR.

This new event looks like it'd make a great addition to Quests. Although I haven't had the time to test it out, I certainly intend to. Until then, I appreciate your effort!

Boosik commented:
Of course you have to test it before merge, I'm already using it on m own server so don't hurry 🙂.
I was looking for a plugin which has this feature and couldn't find any, so I added it myself to the most good looking one.
Although when I saw the code I was bit in a shock that it uses direct field access instead of getters and setters 😀 (what is that used anyway?).

FlyingPikachu commented:
Gave the example quest a try, but got this error shortly after accepting. Timer does not run out.

> [01:15:23] [Server thread/INFO]: HappyPikachu issued server command: /quests take Timed
[01:15:35] [Craft Scheduler Thread - 1/WARN]: [Quests] Plugin Quests v2.10.0-b${env.BUILD_NUMBER} generated an exception while executing task 677
java.lang.IllegalStateException: Not scheduled yet
	at org.bukkit.scheduler.BukkitRunnable.getTaskId(BukkitRunnable.java:134) ~[craftbukkit.jar:git-Spigot-3fb9445-6e3cec8]
	at me.blackvein.quests.ObjectiveTimer.run(ObjectiveTimer.java:25) ~[?:?]
	at org.bukkit.craftbukkit.v1_11_R1.scheduler.CraftTask.run(CraftTask.java:71) ~[craftbukkit.jar:git-Spigot-3fb9445-6e3cec8]
	at org.bukkit.craftbukkit.v1_11_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:52) [craftbukkit.jar:git-Spigot-3fb9445-6e3cec8]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:1.8.0_121]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:1.8.0_121]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_121]

As for the direct field access, I'm afraid it beats me (Blackvein would be behind that).

Boosik commented:
You are right, I'm sorry because I've not tested it again when I added ObjectiveTimer.java:25 line to prevent memory leaks.

I fixed it already. Please try it again with the latest changes.

FlyingPikachu commented:
@Boosik Just tried it again and your example quest works swimmingly!

My only other gripe is how the quest works if I remove the CancelTimer finish-event. Obviously this isn't intended, but let's say a user decides to do it anyway or forgets to include one. Here's what happens:

The quest does not show up in the Quest Journal. If 10 blocks are mined, the "Time left to finish..." messages continue even after the quest has completed. If less than 10 are mined, the quest fails as usual.

Obviously the CancelTimer event is not there to stop the "Time left to finish..." messages if the quest is already completed. This makes me wonder, under what circumstance would you _not_ want to stop the timer? Why not forgo the finish-event and just cancel the timer when it runs out?

Boosik commented:
I did it this way, because I want to be able to create multiple stages under one timer (like horse race with multiple location-to-reach stages - checkpoints) and at the same time guests with timer-per-stage (like fetching some ingredients for a potion the wizard is brewing - one by one with separate timer for each ingredient).
But you gave me an idea how to achieve all the things. Give me a moment.

Boosik commented:
Ok done, when finishing Quest (but not a stage) the timer will be automatically canceled. So you only need to use cancel-timer event when you need to have a separate timers in separate stages.

FlyingPikachu commented:
That makes sense and horse racing sounds like a perfect way to utilize that. I appreciate the changes and plan to merge this later today. Thanks for your hard work!

SXRWahrheit commented:
Thank you for this addition! I have been wanting this feature for years - literally.

Boosik commented:
Well I've been waiting too, but one day I've became sick of waiting and implemented it myself :D
